### PR TITLE
Introduce new metrics schema to kubelet-to-gcm

### PR DESCRIPTION
**Pull request by JacobX33**

### CODE REVIEW

File: kubelet-to-gcm/monitor/kubelet/translate.go [hunk 1]
```diff
@@ -304,114 +411,226 @@ func translateCPU(cpu *stats.CPUStats, tsFactory *timeSeriesFactory, startTime t
...
-	}, startTime, now, diskTotalMD.MetricKind)
-	timeSeries = append(timeSeries, tsFactory.newTimeSeries(resourceLabels, diskTotalMD, diskTotalPoint))
-
-	// Total disk used.
-	diskUsedPoint := tsFactory.newPoint(&v3.TypedValue{
-		Int64Value:      monitor.Int64Ptr(int64(*fs.UsedBytes)),
-		ForceSendFields: []string{"Int64Value"},
-	}, startTime, now, diskUsedMD.MetricKind)
-	timeSeries = append(timeSeries, tsFactory.newTimeSeries(resourceLabels, diskUsedMD, diskUsedPoint))
+	if diskTotalMD, found := metrics["disk_total"]; found {
```
Comment (x13n) [inline kubelet-to-gcm/monitor/kubelet/translate.go]: I don't like these maps with hardcoded keys here. This seems quite error-prone. Can pass the parameters explicitly?

File: kubelet-to-gcm/monitor/kubelet/translate.go [hunk 2]
```diff
@@ -223,20 +303,17 @@ func (t *Translator) translateContainers(pods []stats.PodStats) ([]*v3.TimeSerie
 
 func (t *Translator) translateContainer(podID, namespace string, container stats.ContainerStats, requireFsStats bool) ([]*v3.TimeSeries, error) {
 	var (
-		containerSeries []*v3.TimeSeries
+		containerSeries, memTS, rootfsTS, logfsTS, cpuTS []*v3.TimeSeries
+		err error
 		containerName   = container.Name
+		containerLabels = map[string]string{
+			"namespace": 	namespace,
+			"pod":			podID,
```
Comment (x13n) [inline kubelet-to-gcm/monitor/kubelet/translate.go]: nit: On GitHub the whitespaces here seems misaligned, may need running gofmt.

File: kubelet-to-gcm/monitor/kubelet/source.go [hunk 3]
```diff
@@ -34,8 +34,9 @@ type Source struct {
 
 // NewSource creates a new Source for a kubelet.
 func NewSource(cfg *monitor.SourceConfig) (*Source, error) {
+	metricPrefix := ""
```
[the remark this answers is not in the document]
Comment (JacobX33) [reply kubelet-to-gcm/monitor/kubelet/source.go]: Changed.

File: kubelet-to-gcm/monitor/kubelet/translate.go [hunk 4]
```diff
@@ -28,49 +28,109 @@ import (
...
+	containerMemUsedMD = &metricMetadata{
+		MetricKind: "GAUGE",
+		ValueType:  "INT64",
+		Name:       "kubernetes.io/container/memory/used_bytes",
+	}
+
+	containerPageFaultsMD = &metricMetadata{
+		MetricKind: "DELTA",
+		ValueType:  "INT64",
+		Name:       "kubernetes.io/container/memory/page_fault_kind",
```
[the remark this answers is not in the document]
Comment (JacobX33) [reply kubelet-to-gcm/monitor/kubelet/translate.go]: Done.

File: kubelet-to-gcm/monitor/kubelet/translate.go [hunk 5]
```diff
@@ -28,49 +28,109 @@ import (
...
 		ValueType:  "DOUBLE",
-		Name:       "container.googleapis.com/container/cpu/utilization",
+		Name:       "kubernetes.io/node/cpu/core_usage_time",
 	}
-	diskTotalMD = &metricMetadata{
+
+	nodeMemTotalMD = &metricMetadata{
+		MetricKind: "GAUGE",
+		ValueType:  "INT64",
+		Name:       "kubernetes.io/node/total_bytes",
```
[the remark this answers is not in the document]
Comment (JacobX33) [reply kubelet-to-gcm/monitor/kubelet/translate.go]: Done.

File: kubelet-to-gcm/monitor/kubelet/translate.go [hunk 6]
```diff
@@ -28,49 +28,109 @@ import (
...
+	}
+
+	nodeMemUsedMD = &metricMetadata{
+		MetricKind: "GAUGE",
+		ValueType:  "INT64",
+		Name:       "kubernetes.io/node/used_bytes",
```
[the remark this answers is not in the document]
Comment (JacobX33) [reply kubelet-to-gcm/monitor/kubelet/translate.go]: Done.

File: kubelet-to-gcm/monitor/kubelet/translate.go (hunk 1 above)
Comment (JacobX33) [reply kubelet-to-gcm/monitor/kubelet/translate.go]: Done.

File: kubelet-to-gcm/monitor/kubelet/translate.go (hunk 2 above)
Comment (JacobX33) [reply kubelet-to-gcm/monitor/kubelet/translate.go]: Done.

File: kubelet-to-gcm/monitor/kubelet/translate.go [hunk 7]
```diff
@@ -250,17 +327,29 @@ func (t *Translator) translateContainer(podID, namespace string, container stats
 			ForceSendFields: []string{"DoubleValue"},
 		},
 	}
-	containerSeries = append(containerSeries, tsFactory.newTimeSeries(noLabels, uptimeMD, uptimePoint))
+	metric := containerUptimeMD
+	if t.useOldResourceModel {
```
[the remark this answers is not in the document]
Comment (JacobX33) [reply kubelet-to-gcm/monitor/kubelet/translate.go]: Done.

File: kubelet-to-gcm/monitor/kubelet/translate.go [hunk 8]
```diff
@@ -288,8 +383,20 @@ func (t *Translator) translateContainer(podID, namespace string, container stats
...
+	return translateCPU(cpu, tsFactory, startTime, map[string]*metricMetadata{"usage_time": nodeCpuCoreUsageTimeMD})
+}
+
+func containerTranslateCPU(cpu *stats.CPUStats, tsFactory *timeSeriesFactory, startTime time.Time) ([]*v3.TimeSeries, error) {
+	return translateCPU(cpu, tsFactory, startTime, map[string]*metricMetadata{"usage_time": containerCpuCoreUsageTimeMD})
+}
+
 // translateCPU creates all the TimeSeries for a give CPUStat.
-func translateCPU(cpu *stats.CPUStats, tsFactory *timeSeriesFactory, startTime time.Time) ([]*v3.TimeSeries, error) {
+func translateCPU(cpu *stats.CPUStats, tsFactory *timeSeriesFactory, startTime time.Time, metrics map[string]*metricMetadata) ([]*v3.TimeSeries, error) {
```
[the remark this answers is not in the document]
Comment (JacobX33) [reply kubelet-to-gcm/monitor/kubelet/translate.go]: This got changed.

File: kubelet-to-gcm/monitor/kubelet/translate.go [hunk 9]
```diff
@@ -116,24 +180,24 @@ func (t *Translator) Translate(summary *stats.Summary) (*v3.CreateTimeSeriesRequ
...
-		"cluster_name":   t.cluster,
-		"zone":           t.zone,
-		"instance_id":    t.instanceID,
-		"namespace_id":   "",
-		"pod_id":         "machine",
-		"container_name": "",
+	if t.useOldResourceModel {
+		tsFactory = newTimeSeriesFactory(t.getMonitoredResource(map[string]string{"pod": "machine"}), t.resolution)
+	} else {
+		tsFactory = newTimeSeriesFactory(t.getMonitoredResource(map[string]string{"node": node.NodeName}), t.resolution)
```
[the remark this answers is not in the document]
Comment (JacobX33) [reply kubelet-to-gcm/monitor/kubelet/translate.go]: Done.

File: kubelet-to-gcm/monitor/kubelet/translate.go [hunk 10]
```diff
@@ -304,114 +411,226 @@ func translateCPU(cpu *stats.CPUStats, tsFactory *timeSeriesFactory, startTime t
...
+		resourceLabels["namespace_id"] = labels["namespace"]
+		resourceLabels["pod_id"] = labels["pod"]
+		resourceLabels["container_name"] = labels["container"]
+		return &v3.MonitoredResource{
+			Type:	"gke_container",
+			Labels: resourceLabels,
+		}
+	}
+
+	resourceLabels["location"] = t.zone
```
[the remark this answers is not in the document]
Comment (JacobX33) [reply kubelet-to-gcm/monitor/kubelet/translate.go]: Done.

File: kubelet-to-gcm/monitor/kubelet/translate.go [hunk 11]
```diff
@@ -602,12 +558,14 @@ func (t *Translator) getMonitoredResource(labels map[string]string) *v3.Monitore
...
 	}
 
-	resourceLabels["location"] = t.zone
+	resourceLabels["location"] = t.clusterLocation
 
-	if _, found := labels["node"]; found {
-		resourceLabels["node_name"] = labels["node"]
+	if _, found := labels["container"]; !found {
+		if t.instanceID != "" {
+			resourceLabels["node_name"] = t.instanceID
```
Comment (x13n) [inline kubelet-to-gcm/monitor/kubelet/translate.go]: This should be the node name, not numeric instance id.

File: kubelet-to-gcm/monitor/kubelet/translate.go [hunk 12]
```diff
@@ -617,7 +575,7 @@ func (t *Translator) getMonitoredResource(labels map[string]string) *v3.Monitore
 	resourceLabels["container_name"] = labels["container"]
```
Comment (x13n) [inline kubelet-to-gcm/monitor/kubelet/translate.go]: There should be `pod_name` and `namespace_name`, not id used here.

File: kubelet-to-gcm/monitor/kubelet/translate.go (hunk 2 above)
Comment (x13n) [reply kubelet-to-gcm/monitor/kubelet/translate.go]: It seems this still file needs some reformatting.

File: kubelet-to-gcm/monitor/kubelet/translate.go (hunk 12 above)
Comment (JacobX33) [reply kubelet-to-gcm/monitor/kubelet/translate.go]: Done.

File: kubelet-to-gcm/monitor/kubelet/translate.go (hunk 11 above)
Comment (JacobX33) [reply kubelet-to-gcm/monitor/kubelet/translate.go]: Changed from "instanceID" to "instance".

**Comment by JacobX33** [comment]
Custom label injection ability & testing done.

File: kubelet-to-gcm/monitor/kubelet/translate.go (hunk 11 above)
Comment (x13n) [reply kubelet-to-gcm/monitor/kubelet/translate.go]: This shouldn't be changed, both should be available. "node_name" should use instance name, but "instance_id" label should use the instance id, same as before.

File: kubelet-to-gcm/monitor/main/daemon.go [hunk 13]
```diff
@@ -42,10 +44,13 @@ const (
 )
 
 var (
+	schemaPrefix            = flag.String("schema-prefix", "", "MonitoredResource type prefix, to be appended by 'container', 'pod', and 'node'.")
```
Comment (x13n) [inline kubelet-to-gcm/monitor/main/daemon.go]: This should default to "k8s_".

File: kubelet-to-gcm/monitor/main/daemon.go [hunk 14]
```diff
@@ -42,10 +44,13 @@ const (
...
+	monitoredResourceLabels = flag.String("monitored-resource-labels", "", "Manually specified MonitoredResource labels.")
```
Comment (x13n) [inline kubelet-to-gcm/monitor/main/daemon.go]: Please document the format in README.

File: kubelet-to-gcm/monitor/main/daemon.go (hunk 13 above)
Comment (JacobX33) [reply kubelet-to-gcm/monitor/main/daemon.go]: Done.

File: kubelet-to-gcm/monitor/kubelet/translate.go (hunk 11 above)
Comment (JacobX33) [reply kubelet-to-gcm/monitor/kubelet/translate.go]: "instance_id" label for "gke_container" schema was using instance name before, and that remains unchanged.

File: kubelet-to-gcm/monitor/main/daemon.go (hunk 14 above)
Comment (JacobX33) [reply kubelet-to-gcm/monitor/main/daemon.go]: Done.

**Comment by x13n** [comment]
Thanks, looks good to me now. Can you just bump version tag to 1.3.0 in Makefile?

**Comment by JacobX33** [comment]
Done.

**Comment by x13n** [comment]
Thanks!

**Pull request merged**